lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,961 @@
+            goto Failed; \
+    } \
+} while (0)
+#define COND0(str, lit, pos) (sizeof(lit) - 1 <= (pos) || (str)[pos] == (lit)[pos])
+#define COND1(str, lit, pos) (COND0(str, lit, pos) && COND0(str, lit, pos + 1) && COND0(str, lit, pos + 2))
+#define COND2(str, lit, pos) (COND1(str, lit, pos) && COND1(str, lit, pos + 3) && COND1(str, lit, pos + 6))
+#define COND(str, lit)       (COND2(str, lit, 0)   && COND2(str, lit, 9)       && COND2(str, lit, 18))
+#define CHECK_KEY(lit) ((sizeof(lit) - 1) == keystr_len && COND(keystr, lit))
+
+    khiter_t k;
+    khash_t(ht) *h = mrb_hash_tbl(mrb, env);
+    for (k = kh_begin(h); k != kh_end(h); ++k) {
+        if (!kh_exist(h, k))
+            continue;
+        mrb_value key = h2o_mruby_to_str(mrb, kh_key(h, k));
+        if (mrb->exc != NULL)
+            goto Failed;
+        mrb_value value = kh_value(h, k).v;
+
+        const char *keystr = RSTRING_PTR(key);
+        const mrb_int keystr_len = RSTRING_LEN(key);
+
+        if (CHECK_KEY("CONTENT_LENGTH")) {
+            super->content_length = h2o_strtosize(RSTRING_PTR(value), RSTRING_LEN(value));
+        } else if (CHECK_KEY("HTTP_HOST")) {
+            RETRIEVE_ENV(http_host, 1);
+        } else if (CHECK_KEY("PATH_INFO")) {
+            RETRIEVE_ENV(path_info, 1);
+        } else if (CHECK_KEY("QUERY_STRING")) {
+            RETRIEVE_ENV(query_string, 1);
+        } else if (CHECK_KEY("REMOTE_ADDR")) {
+            RETRIEVE_ENV(remote_addr, 1);
+        } else if (CHECK_KEY("REMOTE_PORT")) {
+            RETRIEVE_ENV(remote_port, 1);
+        } else if (CHECK_KEY("REQUEST_METHOD")) {
+            RETRIEVE_ENV(method, 1);
+        } else if (CHECK_KEY("SCRIPT_NAME")) {
+            RETRIEVE_ENV(script_name, 1);
+        } else if (CHECK_KEY("SERVER_ADDR")) {
+            RETRIEVE_ENV(server_addr, 1);
+        } else if (CHECK_KEY("SERVER_NAME")) {
+            RETRIEVE_ENV(server_name, 1);
+        } else if (CHECK_KEY("SERVER_PORT")) {
+            RETRIEVE_ENV(server_port, 1);
+        } else if (CHECK_KEY("SERVER_PROTOCOL")) {
+            RETRIEVE_ENV(server_protocol, 1);
+        } else if (CHECK_KEY("SERVER_SOFTWARE")) {
+        } else if (CHECK_KEY("h2o.remaining_delegations")) {
+            RETRIEVE_ENV(remaining_delegations, 0);
+        } else if (CHECK_KEY("h2o.remaining_reprocesses")) {
+            RETRIEVE_ENV(remaining_reprocesses, 0);
+        } else if (CHECK_KEY("rack.errors")) {
+            RETRIEVE_ENV(rack_errors, 0);
+        } else if (CHECK_KEY("rack.hijack?")) {
+        } else if (CHECK_KEY("rack.input")) {
+            RETRIEVE_ENV(rack_input, 0);
+        } else if (CHECK_KEY("rack.multiprocess")) {
+        } else if (CHECK_KEY("rack.multithread")) {
+        } else if (CHECK_KEY("rack.run_once")) {
+        } else if (CHECK_KEY("rack.url_scheme")) {
+            RETRIEVE_ENV(scheme, 1);
+        } else if (keystr_len >= 5 && memcmp(keystr, "HTTP_", 5) == 0) {
+            value = h2o_mruby_to_str(mrb, value);
```

Comment:
Maybe we need error check after calling `h2o_mruby_to_str`? I see other places that have the same issue.